doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+      build blocking semantics out of a non-blocking QUIC instance; this is not
+      particularly difficult, though providing custom pollers here would mean
+      there should be no need for an application to do so.)
+
+  - Configuring a socket into non-blocking mode might confuse an application.
+
+    However:
+
+    - Applications will already have to make changes to any network-side BIOs,
+      for example switching from a `BIO_s_socket` to a `BIO_s_dgram`, or from a
+      BIO pair to a `BIO_s_dgram_pair`. Custom BIOs will need to be
+      substantially reworked to switch from bytestream semantics to datagram
+      semantics. Such applications will already need substantial changes, and
+      this is unavoidable.
+
+      Of course, application impacts and migration guidance can (and will) all
+      be documented.
+
+    - In order for an application to be confused by us putting a socket into
+      non-blocking mode, it would need to be trying to use the socket in some
+      way. But it is not possible for an application to pass a socket to our
+      QUIC implementation, and also try to use the socket directly, and have
+      QUIC still work. Using QUIC necessarily requires that an application not
+      also be trying to make use of the same socket.
+
+    - There are some circumstances where an application might want to multiplex
+      other protocols onto the same UDP socket, for example with protocols like
+      RTP/RTCP or STUN; this can be facilitated using the QUIC fixed bit.
+      However, these use cases cannot be supported without explicit assistance
+      from a QUIC implementation and this use case cannot be facilitated by
+      simply sharing a network socket, as incoming datagrams will not be routed
+      correctly. (We may offer some functionality in future to allow this to be
+      coordinated but this is not for MVP.) Thus this also is not a concern.
+      Moreover, it is extremely unlikely that any such applications are using
+      sockets in blocking mode anyway.
+
+Advantages:
+
+  - An application retains full control of its event loop in non-blocking mode.
+
+    When using libssl in application-level blocking mode, via a custom poller
+    interface, the application would actually able to exercise more control over
+    I/O than it actually is at present when using libssl in blocking mode.
+
+  - Feasible to implement and already working in tests.
+    Minimises further development needed to ship.
+
+  - Does not rely on creating threads and can support blocking I/O at the
+    application level without relying on thread assisted mode.
+
+  - Does not require an application-provided network-side custom BIO to be
+    reworked to support concurrent calls to it.
+
+  - Allows performance-optimal implementation of QUIC RFC requirements.
+
+  - Ensures our internal I/O architecture remains flexible for future evolution
+    without breaking compatibility in the future.
```

Comment:
To me this is the most important point. I am really afraid that by supporting blocking I/O we would be cornering ourselves into underperforming and complicated implementation that would not really help our users to take advantage of QUIC properties. Which we would not be able to later get rid of because of the backwards compatibility concerns.